YAML Fragment as alias nodes. #41 (#47)
ietf-wg-httpapi/mediatypes: draft-ietf-httpapi-yaml-mediatypes.md

    - ins: Mike Ralphson
    - ins: Ron Ratovsky
    - ins: Uri Sarid

>   JSON-POINTER: RFC6901

Reviewer: ```suggestion
  JSON-POINTER: RFC6901
  I-D.ietf-jsonpath-base:
```

Fixes the build.

Reviewer: Sorry, I put this in the wrong category -- this should be an informative reference.

The terms "fragment" and "fragment identifier"
in this document are to be interpreted as in {{!URI=RFC3986}}.

The terms "node", "alias node", "anchor" and "named anchor"
in this document are to be intepreded as in [YAML].

## Fragment identification {#application-yaml-fragment}

This section describes how to use
alias nodes (see Section 3.2.2.2 and 7.1 of [YAML])
as fragment identifiers to designate nodes.

A YAML alias node can be represented in a URI fragment identifier
by encoding it into octects using UTF-8 {{!UTF-8=RFC3629}},
while percent-encoding those characters not allowed by the fragment rule
in {{Section 3.5 of URI}}.

If multiple nodes would match a fragment identifier,
the first such match is selected.

A fragment identifier is not guaranteed to reference an existing node.
Therefore, applications SHOULD define how an unresolved alias node
ought to be handled.

Users concerned with interoperability of fragment identifiers:

- SHOULD limit alias nodes to a set of characters
  that do not require encoding
  to be expressed as URI fragment identifiers:
  this is always possible since alias nodes
  and their associated named anchors are a serialization
  detail;

> - SHOULD NOT use alias nodes that matches multiple nodes.

Reviewer: ```suggestion
- SHOULD NOT use alias nodes that match multiple nodes.
```

: HTTP

Fragment identifier considerations:
: An empty fragment identifier references
  the root node.

  A fragment identifier starting with "*"

>   is expressed as a YAML alias node {{application-yaml-fragment}}.

Reviewer: "is expressed"?  Used the wrong way around, ISTM.

"expresses" is a bit weird, how about "stands for"?

(repeated below)

Author: ```suggestion
  is to be interpreted as a YAML alias node {{application-yaml-fragment}}.
```

  A fragment identifier starting with "/"

>   is expressed as a JSON Pointer {{JSON-POINTER}}

Author: ```suggestion
  is to be interpreted as a JSON Pointer {{JSON-POINTER}}
```

  and is evaluated on the YAML representation graph,

>   following alias nodes;

Author: ```suggestion
  resolving alias nodes;
```

  `!mytag` (see Section 2.4 of [YAML]);

~~~ example
 %YAML 1.2
 ---
 non-json-keys:
   0: a number
   2020-01-01: a timestamp
   [0, 1]: a sequence
   ? {k: v}
   : a map
 non-json-value: 2020-01-01
~~~
{: title="Example of mapping keys not supported in JSON" #example-unsupported-keys}

## Fragment identifiers {#int-fragment}

To allow fragment identifiers to traverse alias nodes,
the YAML representation graph needs to be generated before the fragment identifier evaluation.

> It is important that this evaluation will not incurr in the issues mentioned in {{int-yaml-and-json}}

Reviewer: "incurr in"?  (typo, meaning)

Author: ```suggestion
It is important that this evaluation will not cause the issues mentioned in {{int-yaml-and-json}}
```

and in [Security considerations](#security-considerations) such as infinite loops and unexpected code execution.

> Implementers needs to consider that the YAML version and supported features (e.g. merge keys)

Reviewer: ```suggestion
Implementers need to consider that the YAML version and supported features (e.g. merge keys)
```

can impact on the generation of the representation graph (see {{example-merge-keys}}).

In {{application-yaml}}, this document extends the use of specifications based on

> the JSON data model as YAML fragment identifiers.

Reviewer 2: ```suggestion
the JSON data model with support for YAML fragment identifiers.
```
I'm not sure I understand what this sentence is trying to clarify.  Is my suggestion saying what you are trying to say?
